client/client.go

```diff
@@ -12,63 +12,49 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// Package client contains generic representations of clients connecting to different receivers
+// Package client contains generic representations of clients connecting to different receivers. Components,
+// such as processors or exporters, can make use of this information to make decisions related to
+// grouping of batches, tenancy, load balancing, tagging, among others.
 package client // import "go.opentelemetry.io/collector/client"
 
 import (
 	"context"
 	"net"
-	"net/http"
-
-	"google.golang.org/grpc/peer"
 )
 
 type ctxKey struct{}
 
-// Client represents a generic client that sends data to any receiver supported by the OT receiver
-type Client struct {
-	IP string
-}
+// Info contains data related to the clients connecting to receivers.
+type Info struct {
+	// Addr for the client connecting to this collector. Available in a best-effort basis, and generally reliable
+	// for receivers making use of confighttp.ToServer and configgrpc.ToServerOption.
+	Addr net.Addr
 
-// NewContext takes an existing context and derives a new context with the client value stored on it
-func NewContext(ctx context.Context, c *Client) context.Context {
-	return context.WithValue(ctx, ctxKey{}, c)
+	// Auth information from the incoming request as provided by configauth.ServerAuthenticator implementations
+	// tied to the receiver for this connection.
+	Auth AuthData
 }
 
-// FromContext takes a context and returns a Client value from it, if present.
-func FromContext(ctx context.Context) (*Client, bool) {
-	c, ok := ctx.Value(ctxKey{}).(*Client)
-	return c, ok
-}
+// AuthData represents the authentication data as seen by authenticators tied to the receivers.
+type AuthData interface {
+	// GetAttribute returns the value for the given attribute.
+	GetAttribute(string) interface{}
```

Comment:
It will be useful to explain the return value. Do we support non-string auth attributes? If not then maybe return `string`. 

Do we support multiple values for the same attribute (like HTTP headers do)? If we do how is it modelled in the `interface{}`? Is it going to be an array? I do not think multiple values are typically used for auth data, so perhaps restrict to single primitive value per attribute (so a `string` is fine?).

Reply:
It is up to the authenticator, but we have a case for a non-string value already today: the membership information for a given user is a list of strings for the OIDC authenticator. I expanded the godoc with information about the return data type.

Reply:
Can we narrow this to a more specific type instead of `interface{}`? The comment still gives the impression that any value is possible here. If that is the case it is not clear how to use this in the processors. The `Example_processor` doesn't really do anything with the value to make it clear.